slither/utils/upgradeability.py

```diff
@@ -61,11 +64,23 @@
 from slither.tools.read_storage.read_storage import SlotInfo, SlitherReadStorage
 
 
+class TaintedExternalContract(TypedDict):
```

Comment:
I went ahead and made `TaintedExternalContract` a normal class. 

I also added `TaintedFunction` and `TaintedVariable` classes, each of which have a `tainted_by: List[Union[Function, StateVariable]]` property to keep track of which function/variable tainted which. Do you think that's overkill? I could just go back to having `TaintedExternalContract` store a list of Functions and a list of StateVariables, rather than TaintedFunctions and TaintedVariables.

Reply:
I reverted the addition of `TaintedVariable` and `TaintedFunction`, since those mostly just complicated the code. I still have the `TaintedExternalContract` as a normal class, which stores a list of `Function`s and `StateVariable`s.